### PR TITLE
Feature/338 generalized scripter

### LINKED ISSUE #338: scripting engine
Idea of a generic scripting engine, which will autogenerate and run many files
I thought of a more gnerelized version of https://github.com/nbassler/TRiPddd/blob/master/prepare.py

this way one can leave a `shscrip.cfg` in this form:

```
${DIR} = ${NAME}_${BAR}/${X}/.

${NAME} ${FOO} ${BAR} ${X_MIN} ${X_MAX} {X_STEP}
mydir1  1  1  10.0 250.0  10
mydir2  2  4  10.0 250.0  10
mydir3  6  12  10.0 250.0  10
```

### PR DESCRIPTION
**Pull request by nbassler**
Closes #338

### CODE REVIEW

**Comment by nbassler** [comment]
maybe time has come top drop python 2.7?

**Comment by grzanka** [comment]
Most of the tests fails due to the fact that the branch is not following latests master. @nbassler may I rebase ?

File: pymchelper/utils/mcscripter.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+            pass
+
+        if self.symlink:
+            link_file = os.path.join(self.tdir, self.fname)
+            link_target = os.path.join(self.path)
+            link_name = os.path.relpath(link_file, os.path.dirname(link_target))
+
+            try:
+                os.symlink(link_name, link_target)
+            except FileExistsError:
```
Comment (grzanka) [inline pymchelper/utils/mcscripter.py]: It seems python2 doesn't support such Exception

Comment (grzanka) [reply pymchelper/utils/mcscripter.py]: Better to use something low-level like:
```
import os
import errno

try:
    os.mkdir(folderPath)
except OSError as e:
    if e.errno == errno.EEXIST:
        print('Directory not created.')
    else:
        raise
```

from https://stackoverflow.com/a/20790635

File: pymchelper/utils/mcscripter.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    """
+    General MC single file object.
+    This will be used for the template files as well as the generated output files.
+    """
+    def __init__(self):
+        self.fname = ""  # filename
+        self.path = ""   # full path
+        self.lines = []  # list of lines inside this file
+        self.symlink = False  # marker if file is a symlink
+        self.tdir = ""
```
[the remark this answers is not in the document]
Comment (nbassler) [reply pymchelper/utils/mcscripter.py]: will add in comments

File: pymchelper/utils/mcscripter.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+            link_file = os.path.join(self.tdir, self.fname)
+            link_target = os.path.join(self.path)
+            link_name = os.path.relpath(link_file, os.path.dirname(link_target))
+
+            try:
+                os.symlink(link_name, link_target)
+            except FileExistsError:
+                pass
+        else:
+            with open(self.path, 'w') as file:
```
[the remark this answers is not in the document]
Comment (nbassler) [reply pymchelper/utils/mcscripter.py]: yes, of course, I forgot in the haste.

File: pymchelper/utils/mcscripter.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+
+def main(args):
+    """
+    Main function.
+    """
+    logger.setLevel('INFO')
+
+    if len(args) < 1:
+        print("usage: {} <configfile>".format("mcscripter.py"))
```
Comment (grzanka) [inline pymchelper/utils/mcscripter.py]: This won't be valid if you use an executable produced by setup.py. 
(`mcscripter` no `mcscripter.py`)
Consider some simple ArgParse setup, like here
https://github.com/DataMedSci/pymchelper/blob/master/pymchelper/utils/pld2sobp.py#L123

Comment (nbassler) [reply pymchelper/utils/mcscripter.py]: will add it, wanted that anyway

File: pymchelper/utils/mcscripter.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+            else:
+                tf.symlink = False
+            self.files.append(tf)
+
+
+class Generator():
+    """
+    This generates and writes the output files based on the loaded template files and the config file.
+    """
+    def __init__(self, t, cfg):
```
Comment (grzanka) [inline pymchelper/utils/mcscripter.py]: Better to call the parameters `tmpl` or `templ`. It is hard to guess what `t` stands for

File: pymchelper/utils/mcscripter.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    Reading the config file.
+    """
+    def __init__(self, fn):
+        with open(fn) as _f:
+            self.lines = _f.readlines()
+            self.base_dir = os.path.dirname(fn)
+
+            # script must run relative to location of config file.
+            if self.base_dir:
+                os.chdir(self.base_dir)
```
Comment (grzanka) [inline pymchelper/utils/mcscripter.py]: Why a change dir here ? Is it really needed ?

File: pymchelper/utils/mcscripter.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        self.t_dict = {}  # list of table assignments
+
+        keys = []
+        vals = []
+
+        _first_line_in_table = True
+
+        for line in self.lines:
+            # skip comments
+            if line[0] == '#':
```
Comment (grzanka) [inline pymchelper/utils/mcscripter.py]: better: `line.startswith('#')`

File: pymchelper/utils/mcscripter.py (hunk 6 above)
Comment (nbassler) [reply pymchelper/utils/mcscripter.py]: i am afraid so, `os.symlink()`must make relative symlinks, and when they are created outside of the dir, `os.symlink()` will complain that the target do not exist.
Giving paths to files relative to non-base positions will make `os.symlink()` work, but then the links are invalid instead.

I added a few comments in the next commit.

File: pymchelper/utils/mcscripter.py (hunk 5 above)
Comment (nbassler) [reply pymchelper/utils/mcscripter.py]: fixed also the others

File: tests/test_mcscripter.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class TestMcScripter(unittest.TestCase):
+    def test_help(self):
+        """ Print usage and exit normally.
+        """
+        try:
+            pymchelper.utils.mcscripter.main([])
```
Comment (grzanka) [inline tests/test_mcscripter.py]: Why do you expect exit code 0 if no arguments to `mcscripter` are provided ?
With argparse tests will fail. Better to check for exit code 2, or check `--version` option

**Comment by grzanka** [comment]
@nbassler  Appveyor is failing as `os.symlink` is not available for Python 2 on Windows. I'd add there another try-catch block and print an error message if this happends, advising user to use another combination of Python+OS.

**Pull request merged**